### ISSUE TITLE
The more GPU I use, the slower the training speed.

### ISSUE THREAD

**Issue by hobbitlzy** [huggingface/accelerate#192]
I am trying to train the Bert-base-uncased model on Nvidia 3080. However, the strange thing is, the time spent on one step grows sharply with the number of GPU and the total time using multiple GPUs is similar to single GPU.  I directly run the sample code provided on this [link](https://github.com/huggingface/transformers/tree/master/examples/pytorch/language-modeling) and the problem still occurs. BTW, I have run the `transformers.trainer` using multiple GPUs on this machine, and the time per step only increae a little on distributed training. 

The CUDA version shown by `nvidia-smi` is 11.4 and the environment is:
- `transformers` version: 4.11.3
- Platform: Linux-5.11.0-38-generic-x86_64-with-debian-bullseye-sid
- Python version: 3.7.6
- PyTorch version (GPU?): 1.9.0+cu111 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

The relevant outputs on two GPUs are:
```
FutureWarning: The module torch.distributed.launch is deprecated
and will be removed in future. Use torchrun.
Note that --use_env is set by default in torchrun.
If your script expects `--local_rank` argument to be set, please
change it to read from `os.environ['LOCAL_RANK']` instead. See
https://pytorch.org/docs/stable/distributed.html#launch-utility for
further instructions

  FutureWarning,
WARNING:torch.distributed.run:
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed.
*****************************************
cuda:0
10/28/2021 20:21:55 - INFO - __main__ - Distributed environment: MULTI_GPU  Backend: nccl
Num processes: 2
Process index: 0
Local process index: 0
Device: cuda:0
Use FP16 precision: False

cuda:1
10/28/2021 20:21:55 - INFO - __main__ - Distributed environment: MULTI_GPU  Backend: nccl
Num processes: 2
Process index: 1
Local process index: 1
Device: cuda:1
Use FP16 precision: False

..........................

10/28/2021 20:22:28 - INFO - __main__ - ***** Running training *****
10/28/2021 20:22:28 - INFO - __main__ -   Num examples = 4627
10/28/2021 20:22:28 - INFO - __main__ -   Num Epochs = 3
10/28/2021 20:22:28 - INFO - __main__ -   Instantaneous batch size per device = 2
10/28/2021 20:22:28 - INFO - __main__ -   Total train batch size (w. parallel, distributed & accumulation) = 32
10/28/2021 20:22:28 - INFO - __main__ -   Gradient Accumulation steps = 8
10/28/2021 20:22:28 - INFO - __main__ -   Total optimization steps = 435
  0%|▏                                                                                                 | 1/435 [00:11<1:24:51, 11.73s/it]
10/28/2021 20:22:40 - INFO - root - Reducer buckets have been rebuilt in this iteration.
10/28/2021 20:22:40 - INFO - root - Reducer buckets have been rebuilt in this iteration.
 32%|███████████████████████████████▌                                                                  | 140/435 [02:52<05:42,  1.16s/it]
```

The outputs on single GPU:
```
10/28/2021 20:26:47 - INFO - __main__ - Distributed environment: NO
Num processes: 1
Process index: 0
Local process index: 0
Device: cuda
Use FP16 precision: False

.......................

10/28/2021 20:27:49 - INFO - __main__ - ***** Running training *****
10/28/2021 20:27:49 - INFO - __main__ -   Num examples = 4627
10/28/2021 20:27:49 - INFO - __main__ -   Num Epochs = 3
10/28/2021 20:27:49 - INFO - __main__ -   Instantaneous batch size per device = 2
10/28/2021 20:27:49 - INFO - __main__ -   Total train batch size (w. parallel, distributed & accumulation) = 16
10/28/2021 20:27:49 - INFO - __main__ -   Gradient Accumulation steps = 8
10/28/2021 20:27:49 - INFO - __main__ -   Total optimization steps = 870
  4%|███▉                                                                                               | 35/870 [00:17<06:34,  2.12it/s]



```
The hightlight positions are tjat the time per step sharply increase on distributed training and the total time is similar in two settings.

**Comment by hobbitlzy** [comment]
The reason I mention `Tariner` is to rule out the device itself. Moreover, I just use the original `run_mlm_no_trainer.py`  and the given scripts. Actually, I have stucked on this for a week and I am curious whether others met the same problem. I wonder if I set `accelerate config` incorrectly. I post the config file here, although I have checked it many times...

```
compute_environment: LOCAL_MACHINE
deepspeed_config: {}
distributed_type: MULTI_GPU
fp16: false
machine_rank: 0
main_process_ip: null
main_process_port: null
main_training_function: main
num_machines: 1
num_processes: 2

```

**Comment by sgugger** [comment]
Ah, sorry if I misunderstood!

Do you have NVE links between your GPUs? It seems the time lost is all linked to the communication at the end of the `loss.backward()` instruction. There was an old issue on Transformers where they compared iteration times for different number of GPUs, with/without NVE links which showed that the number of GPUs did not necessarily speed up training time overall.

Keep in mind for a fair comparison that you get a 4x bigger batch size with 4 GPUs, so the training is "only" 14/4 = 3.5 times slower

**Comment by hobbitlzy** [comment]
Thanks for your reply. Do you mean this [iuuse](https://github.com/huggingface/transformers/issues/9371#issuecomment-771302472) that mentioned the NVlink? The `nvidia-smi nvlink -s` prints nothing on my device so I guess the GPUs are not connected by NV links. However, I am unsure whether this is the problem. Because the distributed training based on `Trainer` works well. I have made a table to make the picture clearer.
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/lizy/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/lizy/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



  | #GPU | speed | #Iteration | total time | forward/s | backward/s
-- | -- | -- | -- | -- | -- | --
trainer | 1-GPU | 1.26ite/s | 870 | 11:30 | 0.032554 | 0.060441
  | 2-GPU | 1.13ite/s | 435 | 6:20 | 0.032473 | 0.06072
  |   |   |   |   |   |  
accelerator | 1-GPU | 1.25ite/s | 870 | 11:37 | 0.03258 | 0.064011
  | 2-GPU | 1.58s/ite | 435 | 11:26 | 0.032668 | 0.160857
  |   |   |   |   |   |  
DDP | 1-GPU | 1.23ite/s | 870 | 11:44 | 0.03318 | 0.064
  | 2-GPU | 1.56s/ite | 435 | 11:15 | 0.03328 | 0.15845



</body>

</html>

The posted times are a little bit different from previous since I test the codes on another device with 2 v100s. But the trends are similar. The `DDP` means using the built in distributed training of pytorch, which I wrote to debug yesterday. As you can see, the backward time of distributed training using `trainer` does not increase sharply. While using `accelerator` and pytorch built in `DDP` both observe the slow backward.

**Comment by sgugger** [comment]
I was mentioning that issue indeed! Thanks for digging it :-) 

Accelerate and the Trainer use exactly the same code behind the scenes (I wrote both 😅 ): `torch.distributed` so there shouldn't be any differences. You mention DDP, so does that mean you tested your code with both Accelerate and vanilla DDP and got the same slowdown? The one difference I can think of is that the `Trainer` uses `find_unused_parameters=True` when definining the `DistributedDataParallel` model (by default), and I think PyTorch uses the opposite. Could you try?

You can pass along a list of `kwarg_handlers` to the `Accelerator` object to do this with Accelerate (the [DDP one](https://huggingface.co/docs/accelerate/kwargs.html#distributeddataparallelkwargs) here).

**Comment by hobbitlzy** [comment]
These two jobs are great!  Being able to use these two frameworks helps a lot.

>  You mention DDP, so does that mean you tested your code with both Accelerate and vanilla DDP and got the same slowdown? 

Yes, the `Accelerate` observes similar slowdown, as observed using the vanilla DDP. I add this parameter but nothing changes. (BTW, I think setting `find_unused_parameters=True` is supposed be slower🤔). 

I profile the backward time using `torch.profiler` and find an outstanding time called "cudaDeviceSynchronize " (The pink one). Does this noraml? I am wondering whether I wrongly config the `accelerator` so that this happens. I searched the issues and forums but it seems that I am the only one encounters this problem. Are there any traps when using `Accelerator` or `DDP`?

![profiler](https://user-images.githubusercontent.com/35392624/140005093-d6df81ef-40be-4a7d-a147-cffe3d74dd8c.png)

And I attach the accelerator config here:
```
compute_environment: LOCAL_MACHINE
deepspeed_config: {}
distributed_type: MULTI_GPU
fp16: false
machine_rank: 0
main_process_ip: null
main_process_port: null
main_training_function: main
num_machines: 1
num_processes: 2
```

**Comment by sgugger** [comment]
Thanks for investigating! Do you see any difference while using scripts for other tasks? trying to see if there is a specific bug in this MLM script.

**Comment by hobbitlzy** [comment]
I have checked the downstream task using `run_glue_no_trainer.py`, the same slow backward occurs.

**Comment by sgugger** [comment]
Thanks! To make it easier to reproduce, could you post the whole commands you are launching? For instance I'm wondering if you're using gradient accumulation or not.

**Comment by hobbitlzy** [comment]
Sure😁. The lauching script is
```
CUDA_VISIBLE_DEVICES=0,1 accelerate launch --config_file configs/distribute_accelerate_config.yaml test/original_run_mlm_no_trainer.py \
  --dataset_name wikitext \
  --dataset_config_name wikitext-2-raw-v1 \
  --model_name_or_path bert-base-uncased \
  --per_device_train_batch_size 2 \
  --gradient_accumulation_steps 8 \
  --output_dir tmp/test-mlm
```
The `distribute_accelerate_config.yaml`  and the environments are mentioned previously. Thanks for your help!

**Comment by sgugger** [comment]
Ah you *are* using `gradient_accumulation_steps`, that solves the mystery :-)
The Trainer makes sure the gradients are only reduced-averaged when gradient accumulation is finished, something you will have to manually do with Accelerate and DDP.

The forward pass should be done under
```
with model.no_sync():
```
when you don't need to do loss.backward. I will see how we can add a nice API for this in Accelerate.

**Comment by hobbitlzy** [comment]
!!!I take your suggestion and find the total training time shortens to half! 
```diff
    for epoch in range(args.num_train_epochs):
        model.train()
        for step, batch in enumerate(train_dataloader):
-            outputs = model(**batch)
-            loss = outputs.loss
-            loss = loss / args.gradient_accumulation_steps
-            accelerator.backward(loss)
+            if step % args.gradient_accumulation_steps != 0:
+                with model.no_sync():
+                    outputs = model(**batch)
+                    loss = outputs.loss
+                    loss = loss / args.gradient_accumulation_steps
+                    accelerator.backward(loss)
+            else:
+                outputs = model(**batch)
+                loss = outputs.loss
+                loss = loss / args.gradient_accumulation_steps
+                accelerator.backward(loss)

            if step % args.gradient_accumulation_steps == 0 or step == len(train_dataloader) - 1:
                optimizer.step()
                lr_scheduler.step()
                optimizer.zero_grad()
                progress_bar.update(1)
                completed_steps += 1
```

Are they correct? I am not sure so I use the evalution perplexity to check. I change `gradient_accumulation_steps` to make sure the batch size same and set the random seed. But I find the evaluation perplexity are different between single GPU and two GPUs. Does this normal?

**Comment by sgugger** [comment]
Is it sensibly different? The shuffling will be done the same between one or two GPUs, Accelerate makes sure of that, but I'm less sure of the random masking part (which is outside of Accelerate) and any other randomness that might occur.

Glad to see the training speed is better! I'll work on something today and early next week to include something easy in Accelerate. Would love if you could test it when it's ready!

**Comment by hobbitlzy** [comment]
> Is it sensibly different? The shuffling will be done the same between one or two GPUs, Accelerate makes sure of that, but I'm less sure of the random masking part (which is outside of Accelerate) and any other randomness that might occur.
They are slightly different, not a big problem. Although it will be safe if seeing the same result.

> Glad to see the training speed is better! I'll work on something today and early next week to include something easy in Accelerate. Would love if you could test it when it's ready!
Really really really thanks for your patient answer and timely comments. It's my pleasure to try the newest `Accelerate`😁.
